pkg/validation/validate.go

```diff
@@ -122,6 +127,47 @@ func ValidateLabels(limits LabelValidationLimits, userID string, ls []*typesv1.L
 	return nil
 }
 
+type ProfileValidationLimits interface {
+	MaxProfileSizeBytes(userID string) int
+	MaxProfileStacktraceSamples(userID string) int
+	MaxProfileStacktraceSampleLabels(userID string) int
+	MaxProfileStacktraceDepth(userID string) int
+	MaxProfileSymbolValueLength(userID string) int
+}
+
+func ValidateProfile(limits ProfileValidationLimits, userID string, prof *googlev1.Profile, uncompressedSize int, ls phlaremodel.Labels) error {
+	if prof == nil {
+		return nil
+	}
+	if limit := limits.MaxProfileSizeBytes(userID); limit != 0 && uncompressedSize > limit {
+		return NewErrorf(InvalidProfile, ProfileTooBigErrorMsg, phlaremodel.LabelPairsString(ls), uncompressedSize, limit)
+	}
+	if limit, size := limits.MaxProfileStacktraceSamples(userID), len(prof.Sample); limit != 0 && size > limit {
+		return NewErrorf(InvalidProfile, ProfileTooManySamplesErrorMsg, phlaremodel.LabelPairsString(ls), size, limit)
+	}
+	var (
+		depthLimit        = limits.MaxProfileStacktraceDepth(userID)
+		labelsLimit       = limits.MaxProfileStacktraceSampleLabels(userID)
+		symbolLengthLimit = limits.MaxProfileSymbolValueLength(userID)
+	)
+	for _, s := range prof.Sample {
+
+		if depthLimit != 0 && len(s.LocationId) > depthLimit {
+			// truncate the deepest frames
+			s.LocationId = s.LocationId[:depthLimit]
```

Comment:
I just have realised that this truncates parent locations... This may make the flamegraph unreadable. I guess we could achieve better results with cutting the stack tip:
```go
s.LocationId = s.LocationId[len(s.LocationId)-depthLimit:]
```

Reply:
Yeah that wasn't intentional, let's fix it.